ChopChop/ViewModels/Recipes/RecipeFormViewModel.swift

```diff
@@ -0,0 +1,237 @@
+    private var recipeId: Int64?
+    private let storageManager = StorageManager()
+    private var recipeCategoryCancellable = Set<AnyCancellable>()
+
+    private(set) var errorMessage = ""
+    private(set) var isEdit = false
+
+    var pickerSourceType: UIImagePickerController.SourceType = .photoLibrary
+
+    @Published var hasError: Bool = false
+    @Published var isShowingPhotoLibrary = false
+    @Published var image = UIImage()
+    @Published var recipeName = ""
+    @Published var serving = "" {
+        didSet {
+            ensureValidServing()
+        }
+    }
+    @Published var allRecipeCategories = [RecipeCategory]()
+    @Published var recipeCategory = ""
+    @Published var difficulty: Difficulty = .veryEasy
+    @Published var steps = [String]()
+    @Published var ingredients = [RecipeIngredientRowViewModel]()
+    @Published var ingredientParsingString = ""
+    @Published var instructionParsingString = ""
+
+    init(recipe: Recipe) {
+        existingRecipe = recipe
+        recipeId = recipe.id
+        recipeName = recipe.name
+        serving = recipe.servings.description
+        difficulty = recipe.difficulty ?? .veryEasy
+
+        steps = recipe.steps.map({ $0.content })
+        ingredients = recipe.ingredients.map({
+            RecipeIngredientRowViewModel(
+                amount: $0.quantity.value.description,
+                unit: $0.quantity.type,
+                ingredientName: $0.name
+            )
+        })
+        image = storageManager.fetchRecipeImage(name: recipe.name) ?? UIImage()
+        fetchCategories()
+        if let categoryId = recipe.recipeCategoryId {
+            do {
+                recipeCategory = try storageManager.fetchRecipeCategory(id: categoryId)?.name ?? ""
+            } catch {
+
+            }
+        }
+        isEdit = true
+    }
+
+    init() {
+        fetchCategories()
+    }
+
+    private func fetchCategories() {
+        storageManager
+            .recipeCategoriesPublisher()
+            .receive(on: DispatchQueue.main)
+            .sink(receiveCompletion: { [weak self] value in
+                guard let self = self else {
+                    return
+                }
+                switch value {
+                case .failure:
+                    self.allRecipeCategories = []
+                case .finished:
+                    break
+                }
+            },
+            receiveValue: { [weak self] categories in
+                guard let self = self else {
+                    return
+                }
+                self.allRecipeCategories = categories
+            })
+            .store(in: &recipeCategoryCancellable)
+    }
+
+    func ensureValidServing() {
+        let filtered = serving.filter { "0123456789.".contains($0) }
+        if filtered != serving {
+            serving = filtered
+        }
+    }
+
+    func checkFormValid() throws {
+        if recipeName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
```

Comment:
i tot the recipemodel does tis alr

some of the logic here should only be in the models